Update FDA plotting and fitting

Mantid-Matthew opened the pull request:
**Description of work.**

The frequency domain analysis GUI has numerous usability problems. This PR fixes enough of the issues that it is possible to perform transforms on time domain data and then perform fits on the transformed values but there is more work still to be done.

**To test:**
 * Test that you can perform frequency domain transforms and perform fits on single data workspaces. Check that the plots update in the expected manner when you do so. Get in touch with me to explain the workflows in this GUI is you are not already familiar with it.

<!-- Instructions for testing. -->

Fixes 28742 <!-- and fix #xxxx or close #xxxx xor resolves #xxxx -->
<!-- alternative
*There is no associated issue.*
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
-->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Mantid-Matthew commented on scripts/Muon/GUI/Common/plot_widget/plot_widget_view.py:
```diff
@@ -120,3 +138,41 @@ def set_plot_type(self, plot_type: str):
+        """
+        self.tiled_plot_checkbox.setEnabled(True)
+        self.tiled_by_combo.setEnabled(True)
+        self.plot_raw_checkbox.setEnabled(True)
```
Good point

Mantid-Matthew commented:
> 
> 
> This is a nice set of changes for both GUIs, it was a good idea to introduce PlotMode. Just a few comments and one issue:
> 
> For me the FDA fitting isn't working, it complains about there not being a fit function defined. I think this is partly due to me. To support two tabs accessing the same model and allow any row of the sequential table to be fitting individually, I created a workspace fit function [map](https://github.com/mantidproject/mantid/blob/359994311ec8c708a0c2406b70de9d8d243e5706/scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py#L444). I think its not creating this workspace map correctly for FDA, as the keys of the map are:
> 
> `[frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'}), frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'}), frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'})]`
> 
> Whereas the input workspace to the fit is 'FFT; Re MUSR62260; Group; fwd; Asymmetry; FD; Im MUSR62260; Group; fwd; Asymmetry; FD_mod', which means it cannot find the correct function (as it usees this workspace name to grab the correct fit function).

Yes you are entirely right here. I've now added a fix which I think will get the fitting to work for the freuency domain analysis. I suspect this will need to be further modified if batch fitting is ever added to the FDA GUI. But as that may not ever happen and I am not quite sure of what form it will take decided not to complicate things further for now.

Mantid-Matthew commented:
@StephenSmith25 Just remembered that I was also planning to disable the Ts Asymmetry mode option in the FDA GUI so hold off on looking at this again until I have done that.

StephenSmith25 commented:
This works well now, just one thing which you may want to change. The `Time Start` and `Time end` boxes in the fitting tab should probably be renamed to something more appropriate for FDA.

Mantid-Matthew commented:
> This works well now, just one thing which you may want to change. The `Time Start` and `Time end` boxes in the fitting tab should probably be renamed to something more appropriate for FDA.

Yes good spot and given some more appropriate initial values as well.

AnthonyLim23 commented on scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_view.py:
```diff
@@ -35,6 +35,11 @@ def __init__(self, parent=None):
 
+        if is_frequency_domain:
+            self.hide_simultaneous_fit_options()
+            self.fit_options_table.hideRow(3)
```
what are these magic numbers?

Mantid-Matthew replied:
These are refencing the names of the table displaying information. I've changed them to named constants to make this clearer.

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:
```diff
@@ -11,16 +11,21 @@
 from mantid.dataobjects import Workspace2D
 
+MUON_ANALYSIS_DEFAULT_X_RANGE = [0.0, 15.0]
+FREQUENCY_DOMAIN_ANALYSIS_DEFAULT_X_RANGE = [0.0, 1000.0]
```
should we get these from the context?

Mantid-Matthew replied:
Yes good idea that removes the need for one of the if statements.

AnthonyLim23 commented on scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:
```diff
@@ -83,7 +94,73 @@ def handle_data_updated(self):
+            )
+
+    def handle_plot_mode_changed(self, plot_mode : PlotMode):
+        if isinstance(self.context, FrequencyDomainAnalysisContext):
```
I have seen this a few times. Would it be worth making it a method of the context?

Mantid-Matthew replied:
I don't think you gain much by adding a method to the context identifying what type of context it is when python as a language allows you to easily check the type of objects. The ideally we should be redeucing the number of these checks going forward as well as we differentiate between the different GUI's by more clearly seperating there code into different classes.

AnthonyLim23 commented:
clicking `calculate FFT` causes an error `QComboBox object has no attribute setCurrentText`

AnthonyLim23 commented:
the xrange doesnt seem to change to frequency when plotting it

AnthonyLim23 commented:
the plots shown seem to be somewhat random. If you do a default FFT on MUSR62260, then change the Im group to something else. The new FFT will be plotted on only some plots.

AnthonyLim23 commented:
doing a fit in the time domain GUI throws a long stack trace. Seems to be from the ADS observer and might be because I have loaded the same data into FDA first.

Mantid-Matthew commented:
> 
> 
> the xrange doesnt seem to change to frequency when plotting it outside of fit

Good spot, I was planning to remove the ability to plot frequency outside of fits but forgot

Mantid-Matthew commented:
> 
> 
> clicking `calculate FFT` causes an error `QComboBox object has no attribute setCurrentText`

Looks like setCurrentText became deprecated at some point and has presumably dissapeared on the version of Qt that you are using. I'll use something else. Out of interest which system/qt version are you testing on?

Mantid-Matthew commented:
> 
> 
> the plots shown seem to be somewhat random. If you do a default FFT on MUSR62260, then change the Im group to something else. The new FFT will be plotted on only some plots.

Could you elaborate on how to do this? For me it always plots the mod of the latest FFT done.

StephenSmith25 commented:
> doing a fit in the time domain GUI throws a long stack trace. Seems to be from the ADS observer and might be because I have loaded the same data into FDA first.

I've noticed this as well. If you've had both GUIS open the ADS observers seem to conflict. Think this needs to be looked into.

Mantid-Matthew commented:
> 
> 
> doing a fit in the time domain GUI throws a long stack trace. Seems to be from the ADS observer and might be because I have loaded the same data into FDA first.

I am struggling to reproduce this. Can you remember the precise order of steps which produced the stacktrace? Also could you send me the stacktrace.

AnthonyLim23 commented:
> > clicking `calculate FFT` causes an error `QComboBox object has no attribute setCurrentText`
> 
> Looks like setCurrentText became deprecated at some point and has presumably dissapeared on the version of Qt that you are using. I'll use something else. Out of interest which system/qt version are you testing on?

I think its Qt4, Im using MantidPlot on Windows (get an error trying to run workbench thats unrelated)

AnthonyLim23 commented:
> > the xrange doesnt seem to change to frequency when plotting it outside of fit
> 
> Good spot, I was planning to remove the ability to plot frequency outside of fits but forgot. Do we want to disallow looking at frequency in data mode or add in the scale chaneing?

we want transform tab to show freq data. But can do that by switching to data

Mantid-Matthew commented:
In summary I have now:
  * Corrected the issue with it not working in qt4
  * Removed the ability to perform frequency domain plots from the data plots for now
  * The issue with plots not updating when performing new transformed is solved by point 1 here.
  * We've agreed to leave the ADS observer issue for another PR

So should be good for a re-look

AnthonyLim23 commented:
flake8 errors

AnthonyLim23 commented:
Do an FFT and the fit to the data.
Go to the results tab and it has been added
Go back to fit tab
Fit some different data 
Go to the results tab, the data has not been added

Mantid-Matthew commented:
> Do an FFT and the fit to the data.
> Go to the results tab and it has been added
> Go back to fit tab
> Fit some different data
> Go to the results tab, the data has not been added

This issues predates this PR so as this has gotten quite large I have added a new issue for this #29015

AnthonyLim23 commented:
@mantidproject/gatekeepers could we please merge this?
